### PR TITLE
Fix #124276 batch markdown file requests

### LINKED ISSUE #124276: Revisit markdown having to scan over all markdown files in workspace for open symbol
Extracted from https://github.com/microsoft/vscode/issues/124176

I find it questionable that we need to read all markdown files from the entire workspace in parallel. We should consider to add a maximum cap and also possibly debounce this and not just join all promises in parallel:

https://github.com/microsoft/vscode/blob/d5f4e119e058e1d01ec4f994f757703348c37db8/extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts#L29-L33

### PR DESCRIPTION
**Pull request by nrayburn-tech**
- I modified the function to batch the requests, so that they are done in batches of 20.  This is a bit less efficient than pooling the request, but I didn't think this really needed the extra complexity.  I can use a pool instead, if it really matters.
- Also modified a function return type to remove `| undefined`.  This doesn't seem like a valid return type for that function and allowed some ifs to be removed, and a filter on the batching requests function.

I would expect using batches to reduce performance on workspaces that have a large number of `.md` files.  I don't have a good way to test this for comparison though.

This PR fixes #124276.

### CODE REVIEW

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 1]
```diff
@@ -26,10 +26,38 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
+	 */
+	async getAllMarkdownDocuments(): Promise<SkinnyTextDocument[]> {
+		const maxConcurrent = 20;
+		const ignoreList: (string | vscode.RelativePattern)[] = ['**/node_modules/**'];
+		const docList: SkinnyTextDocument[] = [];
+
+		let isDone = false;
+		while (!isDone) {
+			await new Promise(async (resolve) => {
+				const resources = await vscode.workspace.findFiles('**/*.md', `{${ignoreList.join(',')}}`, maxConcurrent);
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:43]: Not sure about using the ignore list this way since it will get very large for large workspaces. For the batching, how about executing `findFiles` once and then batching the reads themselves?

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 2]
```diff
@@ -26,10 +26,38 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
+				if (resources.length < maxConcurrent) {
+					isDone = true;
+				}
+
+				await Promise.all(resources.map(doc => {
+					// Convert path to be relative from the workspace
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:49]: Why is this  converationneeded?

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 3]
```diff
@@ -56,16 +84,12 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
 
 		this._watcher.onDidChange(async resource => {
 			const document = await this.getMarkdownDocument(resource);
-			if (document) {
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:59]: Can you please make these changes in a separate PR? I'd like to keep this one scoped to just batching the requests

**Review by mjbvz — CHANGES_REQUESTED** [review]

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts (hunk 1 above)
Comment (nrayburn-tech) [reply extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:43]: Just to clarify, the issue isn't that we are calling `workspace.findFiles('**/*.md')`, it is that we are taking that result and calling `getMarkdownDocument` for all `.md` files in the workspace at a single time.
Is that right?
If so, then I agree with you that this should be changed and an ignore list shouldn't be used like this.  I thought that we should be avoiding calling `findFiles` for all `.md` files at once and this was the only way to do it.

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts (hunk 2 above)
Comment (nrayburn-tech) [reply extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:49]: `workspace.findFiles` is relative to the workspace, so I couldn't just turn around and pass back in `doc.fsPath` as an exclude.
`/Users/nrayburn/documents/vscode/contributing.md` vs `contributing.md` for example.  Maybe there is a better way to do this that I missed, but this was the only way I could find to get it working.

This becomes irrelevant, if we only do the `workspace.findFiles` call once though.

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts (hunk 1 above)
Comment (mjbvz) [reply extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:43]: Yes `findFiles` is safe. The problem is that we then try to open every found file in parallel (which could be many thousands of files)

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 4]
```diff
@@ -26,10 +26,26 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
+	 */
+	async getAllMarkdownDocuments(): Promise<SkinnyTextDocument[]> {
+		const maxConcurrent = 20;
+		const docList: SkinnyTextDocument[] = [];
 		const resources = await vscode.workspace.findFiles('**/*.md', '**/node_modules/**');
-		const docs = await Promise.all(resources.map(doc => this.getMarkdownDocument(doc)));
-		return docs.filter(doc => !!doc) as SkinnyTextDocument[];
+
+		// add 1 in case of any remaining files,
+		// ex (10 % 3) = 3, needs 1 more loop to complete
```
Comment (nrayburn-tech) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:41]: Is this comment necessary?

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 5]
```diff
@@ -26,10 +26,26 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
+		for (let i = 0; i < resources.length % maxConcurrent + 1; i++) {
+			const resourceBatch = resources.slice(i * maxConcurrent, (i + 1) * maxConcurrent);
+			(await Promise.all(resourceBatch.map(async doc => {
+				return await this.getMarkdownDocument(doc);
+			}))).filter((doc) => !!doc);
```
Comment (nrayburn-tech) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:46]: Should this be broken up for readability?

**Review by nrayburn-tech — COMMENTED** [review]
@mjbvz updated.

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts (hunk 1 above)
Comment (james7618891) [reply extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts]: > Just to clarify, the issue isn't that we are calling `workspace.findFiles('**/*.md')`, it is that we are taking that result and calling `getMarkdownDocument` for all `.md` files in the workspace at a single time.
> 
> Is that right?
> 
> If so, then I agree with you that this should be changed and an ignore list shouldn't be used like this.  I thought that we should be avoiding calling `findFiles` for all `.md` files at once and this was the only way to do it.

> ###- [ ] - - - - - - - [x] - - - - - - - - - - - - - Uat zos lob ndrangt ndangt ndrangt @nrayburn-tech [_![image](https://user-images.githubusercontent.com/84305127/119748532-aa4eea80-```

```bec7-11eb-8e6e-6fa2deea02ae.png)_]()

**Review by james7618891 — COMMENTED** [review]
Visa

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 6]
```diff
@@ -26,10 +26,26 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
-	async getAllMarkdownDocuments() {
+	/**
+	 * Reads and parses all .md documents in the workspace.
+	 * Files are processed in batches, to keep the number of open files small.
+	 *
+	 * @returns Array of processed .md files.
+	 */
+	async getAllMarkdownDocuments(): Promise<SkinnyTextDocument[]> {
+		const maxConcurrent = 20;
+		const docList: SkinnyTextDocument[] = [];
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:37]: Looks like `docList` is never modified?

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 7]
```diff
@@ -26,10 +26,26 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
 		const resources = await vscode.workspace.findFiles('**/*.md', '**/node_modules/**');
-		const docs = await Promise.all(resources.map(doc => this.getMarkdownDocument(doc)));
-		return docs.filter(doc => !!doc) as SkinnyTextDocument[];
+
+		// add 1 in case of any remaining files,
+		// ex (10 % 3) = 3, needs 1 more loop to complete
+		for (let i = 0; i < resources.length % maxConcurrent + 1; i++) {
+			const resourceBatch = resources.slice(i * maxConcurrent, (i + 1) * maxConcurrent);
+			(await Promise.all(resourceBatch.map(async doc => {
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:44]: [nit] Just return `this.getMarkdownDocument` so you don't need `async` or `await`

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 8]
```diff
@@ -26,10 +26,26 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
+	async getAllMarkdownDocuments(): Promise<SkinnyTextDocument[]> {
+		const maxConcurrent = 20;
+		const docList: SkinnyTextDocument[] = [];
 		const resources = await vscode.workspace.findFiles('**/*.md', '**/node_modules/**');
-		const docs = await Promise.all(resources.map(doc => this.getMarkdownDocument(doc)));
-		return docs.filter(doc => !!doc) as SkinnyTextDocument[];
+
+		// add 1 in case of any remaining files,
+		// ex (10 % 3) = 3, needs 1 more loop to complete
+		for (let i = 0; i < resources.length % maxConcurrent + 1; i++) {
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:42]: Instead of using `%`, could you do something like: `for (let i = 0; I < resource.length; i += maxConcurrent)`

**Review by mjbvz — CHANGES_REQUESTED** [review]

**Comment by nrayburn-tech** [comment]
@mjbvz should all be resolved now.

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 9]
```diff
@@ -26,10 +26,23 @@ class VSCodeWorkspaceMarkdownDocumentProvider extends Disposable implements Work
...
+		const maxConcurrent = 20;
+		let docList: SkinnyTextDocument[] = [];
 		const resources = await vscode.workspace.findFiles('**/*.md', '**/node_modules/**');
-		const docs = await Promise.all(resources.map(doc => this.getMarkdownDocument(doc)));
-		return docs.filter(doc => !!doc) as SkinnyTextDocument[];
+
+		for (let i = 0; i < resources.length; i += maxConcurrent) {
+			const resourceBatch = resources.slice(i, i + maxConcurrent);
+			const documentBatch = (await Promise.all(resourceBatch.map(this.getMarkdownDocument))).filter((doc) => !!doc) as SkinnyTextDocument[];
+			docList = [...docList, ...documentBatch];
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts:43]: [nit] Use `.push(...documentBatch)` here instead of creating a new array each time

**Review by mjbvz — COMMENTED** [review]
Very small change but otherwise looks good for merging

**Comment by nrayburn-tech** [comment]
@mjbvz updated with the change and merged upstream.

**Comment by mjbvz** [comment]
Thanks! This is schooled to go out with VS Code 1.60 at the end of this month
